File: org.eclipse.scout.rt.shared/src/main/java/org/eclipse/scout/rt/shared/services/common/code/CODES.java
/*******************************************************************************
 * Copyright (c) 2010-2016 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 ******************************************************************************/
package org.eclipse.scout.rt.shared.services.common.code;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

import org.eclipse.scout.rt.platform.BEANS;
import org.eclipse.scout.rt.platform.util.CollectionUtility;

/**
 * Convenience accessor for service ICodeService
 */
public final class CODES {

  private CODES() {
  }

  /**
   * @param id
   * @return Note that this method does not load code types, but only searches code types already loaded into the code
   *         service using {@link #getAllCodeTypes(String)}, {@link #getCodeType(Class)} etc.
   */
  public static <T> ICodeType<T, ?> findCodeTypeById(T id) {
    return BEANS.get(ICodeService.class).findCodeTypeById(id);
  }

  public static List<ICodeType<?, ?>> getCodeTypes(Class<?>... types) {
    List<Class<? extends ICodeType<?, ?>>> typeList = varargToList(types);
    return BEANS.get(ICodeService.class).getCodeTypes(typeList);
  }

  public static List<ICodeType<?, ?>> getCodeTypes(List<Class<? extends ICodeType<?, ?>>> types) {
    return BEANS.get(ICodeService.class).getCodeTypes(types);
  }

  public static <CODE_ID_TYPE, CODE extends ICode<CODE_ID_TYPE>> CODE getCode(Class<CODE> type) {
    return BEANS.get(ICodeService.class).getCode(type);
  }

  public static <T extends ICodeType> T reloadCodeType(Class<T> type) {
    return BEANS.get(ICodeService.class).reloadCodeType(type);
  }

  public static List<ICodeType<?, ?>> reloadCodeTypes(Class<?>... types) {
    List<Class<? extends ICodeType<?, ?>>> typeList = varargToList(types);
    return BEANS.get(ICodeService.class).reloadCodeTypes(typeList);
  }

  public static List<ICodeType<?, ?>> reloadCodeTypes(List<Class<? extends ICodeType<?, ?>>> types) {
    return BEANS.get(ICodeService.class).reloadCodeTypes(types);
  }

  public static void invalidateCodeType(Class<? extends ICodeType> type) {
    BEANS.get(ICodeService.class).invalidateCodeType(type);
  }

  public static void invalidateCodeTypes(Class<?>... types) {
    List<Class<? extends ICodeType<?, ?>>> typeList = varargToList(types);
    BEANS.get(ICodeService.class).invalidateCodeTypes(typeList);
  }

  public static void invalidateCodeTypes(List<Class<? extends ICodeType<?, ?>>> types) {
    BEANS.get(ICodeService.class).invalidateCodeTypes(types);
  }

  /**
   * Converts the given classes vararg into a {@link List}.
   */
  @SuppressWarnings("unchecked")
  private static List<Class<? extends ICodeType<?, ?>>> varargToList(Class<?>... types) {
    if (types == null) {
      return CollectionUtility.emptyArrayList();
    }
    List<Class<? extends ICodeType<?, ?>>> typeList = new ArrayList<>(types.length);
    for (Class<?> t : types) {
      if (ICodeType.class.isAssignableFrom(t)) {
        typeList.add((Class<? extends ICodeType<?, ?>>) t);
      }
    }
    return typeList;
  }

  public static Collection<ICodeType<?, ?>> getAllCodeTypes(String classPrefix) {
    Set<Class<? extends ICodeType<?, ?>>> allCodeTypeClasses = getAllCodeTypeClasses(classPrefix);
    List<Class<? extends ICodeType<?, ?>>> list = CollectionUtility.arrayList(allCodeTypeClasses);
    return getCodeTypes(list);
  }

  public static Set<Class<? extends ICodeType<?, ?>>> getAllCodeTypeClasses(String classPrefix) {
    final Set<Class<? extends ICodeType<?, ?>>> filteredClasses = new LinkedHashSet<>();
    final Collection<Class<? extends ICodeType<?, ?>>> classes = BEANS.get(ICodeService.class).getAllCodeTypeClasses();
    for (Class<? extends ICodeType<?, ?>> c : classes) {
      if (c.getName().startsWith(classPrefix)) {
        filteredClasses.add(c);
      }
    }
    return filteredClasses;
  }
}

File: eclipse-scout-core/src/code/codes.js
/*
 * Copyright (c) 2014-2018 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 */
import {texts} from '../index';
import {objects} from '../index';
import {CodeType} from '../index';
import * as $ from 'jquery';
import {arrays} from '../index';
import {scout} from '../index';


/**
 * This default language is used whenever a code registers its texts in scout.texts.
 */
let defaultLanguage = 'en';

let registry = {};

export function bootstrap(url) {
  var promise = url ? $.ajaxJson(url) : $.resolvedPromise({});
  return promise.then(_preInit.bind(this, url));
}

//private
export function _preInit(url, data) {
  if (data && data.error) {
    // The result may contain a json error (e.g. session timeout) -> abort processing
    throw {
      error: data.error,
      url: url
    };
  }
  init(data);
}

export function init(data) {
  data = data || {};
  Object.keys(data).forEach(function(codeTypeId) {
    add(data[codeTypeId]);
  }, this);
}

/**
 * @param codes one or more codeTypes, maybe an object or an array
 */
export function add(codeTypes) {
  codeTypes = arrays.ensure(codeTypes);
  codeTypes.forEach(function(codeType) {
    codeType = CodeType.ensure(codeType);
    registry[codeType.id] = codeType;
  }, this);
}

/**
 * @param codeTypes code types or code type ids to remove
 */
export function remove(codeTypes) {
  codeTypes = arrays.ensure(codeTypes);
  codeTypes.forEach(function(codeType) {
    var id;
    if (typeof codeType === 'string') {
      id = codeType;
    } else {
      id = codeType.id;
    }
    delete registry[id];
  }, this);
}

/**
 * Returns a code for the given codeId. When you work with hard-coded codes
 * you should always use this function and not <code>optGet</code>.
 *
 * The codeId is a string in the following format:
 *
 * "[CodeType.id] [Code.id]"
 *
 * Examples:
 * "71074 104860"
 * "MessageChannel Phone"
 *
 * CodeType.id and Code.id are separated by a space.
 * The Code.id alone is not unique, that's why the CodeType.id must be always provided.
 *
 * You can also call this function with two arguments. In that case the first argument
 * is the codeTypeId and the second is the codeId.
 *
 * @param {string} vararg either only "[CodeType.id]" or "[CodeType.id] [Code.id]"
 * @param {string} [codeId]
 * @returns {Code} a code for the given codeId
 * @throw {Error} if code does not exist
 */
export function get(vararg, codeId) {
  return _get('get', objects.argumentsToArray(arguments));
}

/**
 * Same as <code>get</code>, but does not throw an error if the code does not exist.
 * You should always use this function when you work with codes coming from a dynamic data source.
 *
 * @param vararg
 * @param codeId
 * @returns {Code} code for the given codeId or undefined if code does not exist
 */
export function optGet(vararg, codeId) {
  return _get('optGet', objects.argumentsToArray(arguments));
}

//private
export function _get(funcName, funcArgs) {
  var codeTypeId, codeId;
  if (funcArgs.length === 2) {
    codeTypeId = funcArgs[0];
    codeId = funcArgs[1];
  } else {
    var tmp = funcArgs[0].split(' ');
    if (tmp.length !== 2) {
      throw new Error('Invalid string. Must have format "[CodeType.id] [Code.id]"');
    }
    codeTypeId = tmp[0];
    codeId = tmp[1];
  }
  scout.assertParameter('codeTypeId', codeTypeId);
  scout.assertParameter('codeId', codeId);
  return codeType(codeTypeId)[funcName](codeId);
}

export function codeType(codeTypeId, optional) {
  var codeType = registry[codeTypeId];
  if (!optional && !codeType) {
    throw new Error('No CodeType found for id=' + codeTypeId);
  }
  return codeType;
}

export function generateTextKey(code) {
  // Use __ as prefix to reduce the possibility of overriding 'real' keys
  return '__code.' + code.id;
}

/**
 * Registers texts for a code. It uses the method generateTextKey to generate the text key.
 * The texts for the default locale specified by defaultLanguage are used as default texts.
 *
 * @param code the code to register the text for
 * @param textsArg an object with the languageTag as key and the translated text as value
 * @return the generated text key
 */
export function registerTexts(code, textsArg) {
  var key = generateTextKey(code);

  // In case of changed defaultLanguage clear the 'default' entry
  texts.get('default').remove(key);

  for (var languageTag in textsArg) { // NOSONAR
    var text = textsArg[languageTag];
    // Use defaultLanguage as default, if specified (may be changed or set to null by the app).
    if (languageTag && languageTag === defaultLanguage) {
      languageTag = 'default';
    }
    texts.get(languageTag).add(key, text);
  }
  return key;
}


export default {
  add,
  bootstrap,
  codeType,
  defaultLanguage,
  generateTextKey,
  get,
  init,
  optGet,
  registerTexts,
  registry,
  remove
};
